Rewrite RGB21 spec using Contractum & concept of interfaces (#137)
LNP-BP/LNPBPs: lnpbp-0021.md

```
LNPBP: 0021
Vertical: Smart contracts
Title: RGB non-fungible assets interface for collectibles (RGB-21)
Authors: Dr Maxim Orlovsky <orlovsky@lnp-bp.org>,

>          Hunter Trello,

Reviewer: ```suggestion
         Hunter Trujillo,
```
Haha, [Trello](https://trello.com) is something different :)

         Federico Tenga,
         Carlos Roldan,

>          Olgsa Ukolova,

Reviewer: ```suggestion
         Olga Ukolova,
```

## Abstract

## Background

## Motivation

## Design

- [x] Media as URI (which can be attachments distributed with Storm or usual URLs)
- [x] Small media ("previews" if given together with large media)
- [x] Fraction of assets
- [x] Engravings
- [x] Reserves
- [x] Possible decentralized issue

## Specification

Interface specification is the following Contractum code:

```haskell
-- # Defining main data structures

-- collectibles are usually scarse, so we limit their max number to 64k
data ItemsCount :: U16

-- each collectible item is unique and must have an id
data TokenId :: U16

-- Collectibles are owned in fractions of a single item; here the owned
-- fraction means `1/F`. Ownership of the full item is specified `F=1`;
-- half of an item as `F=2` etc.
data OwnedFraction :: U64

data TxOut :: txid [Byte ^ 32], vout U16

-- allocation of a single token or its fraction to some transaction output
data Allocation :: TokenId, OwnedFraction

data Media ::
    type MimeType,
    data [Byte]

data Attachment ::
    type MimeType,
    uri: Uri

data POR :: -- proof of reserves
    reserves TxOut,
    proof [Byte] -- schema-specific proof 

data Token ::

>     id Id,

Author: ```suggestion
    id TokenId,
```

    name [Ascii ^ 1..40],
    details [Unicode ^ 40..256]?,

>     media Media?,
>     attachment Attachment?,

Author: ```suggestion
    preview Media?, -- always embedded preview media < 64kb
    media Attachment?, -- external media which is the main media for the token
    attachments { U8 ^ ..20 -> Attachment } -- auxillary attachments by type (up to 20 attachments)
```

Reviewer: I don't suppose we could do something like 255 attachments? For example, if this is used for songs, instead of having an RGB21 for each song, we could have an RGB21 for each album, and albums rarely have more than 20 songs, and certainly fewer than 100. Not sure if there's a threshold where byte space efficiency makes sense to do 255 instead of say 100.

Also bear in mind, it's spelled `auxiliary`, not two ls, but an l and a i. With American pronunciation, it's said differently than it's spelled. I ordinarily wouldn't be so pedantic about spelling, but... Such a mistake might be harder to correct in the future, who knows.

Author: These attachments are not individual non-fungible tokens, but attachments to a single token, each of which have a unique MIME type. I.e. in case of ablum, the album is an contract under a scheme implementing RGB21, with each song being a `Token` here, which `media` is the audio recording, and `attachments` are lyrics, and maybe something else (cover image?). I doubt we need a lot of attachments, but I agree that 20 is arbitrary small amount.

Reviewer: Right, I'm suggesting the limit should be higher so the token would be the album and not each individual song.

Author: Hm, and what would be contract in that case?

Making token the album will:
- disallow selling/owning songs by one
- leave them without names
- leave themselves without attachments (like do not providing lyrics)

At the same time it is not clear which advantages we would get...

Reviewer: An album associates them. There is a way around this, and it's to have a `Parent?`, where there's a token for an album and each song, and each song specifies the RGB ID for the parent album, which would have no parent, only an artist, which is different since that's the issuer, though maybe it could point to an RGB22.

Author: Sorry, I am still not getting why we need this complexity if an RGB21 _contract_ can represent an album, and a token inside the contract a song.

Reviewer: I see, so an album contract would have a token for each song... That's great, I like it! Each token will have a unique id, correct? To make them enumerable?

Author: Correct. That what I was trying to propose

Reviewer: Sounds good. We can consider this point resolved, then.

    reserves POR? -- output containing locked bitcoins; how reserves are
                  -- proved is a matter of a specific schema implementation

data Nomination :: 
    ticker [Ascii ^ 1..8],
    name [Ascii ^ 1..40],
    details [Unicode ^ 40..256]?,
    contract [Unicode]??,

> 

Author: ```suggestion

data Auxillary ::
    -- list of allowed sources
    sources [Source]
    
     -- each attachment type is a mapping from attachment id 
     -- (used as `Token.attachments` keys) to a short Ascii string
     -- verbally explaining the type of the attachment for the UI
     -- (like "sample" etc).
    attTypes { U8 ^ ..32 -> [Ascii ^ 1..20] }

data Source :: url(Proto, Dns) | urn(UrnPrefix) | storm(NodeAddr)
data Proto :: http | https | httpxk | ws | wss | wssxk

```

Reviewer: `Auxillary` -> `Auxiliary`
If it costs no more space, let's call it `attachmentTypes` instead of `attTypes`, since it could also be mistaken for the word `attribute`

Also, I'm not sure the idea behind separating the source from the identifier is sufficiently captured here. If there were room for an optional argument in each, such as:
`data Source :: url(Proto, Dns, Id?) | urn(UrnPrefix, Id?) | storm(NodeAddr, Id?)`
and then the Id could be provided as a separate field, if a lookup failed on one source, it could be tried on another. That's why I think there should be an array of sources capable of being provided.

Reviewer: Also, would it not be simpler to just have a `attachmentType` on each `Attachment` than have a separate lookup?
Additionally, having auxiliary attachments might be overcomplicating it. I'm not sure it necessarily makes sense to have a single special "primary" attachment, especially if a convention were used that if such a thing were desired, the first Attachment in the array (index 0) would be used.

Author: The idea is that an NFT contract defines which attachment types can be out there, like for an album they would be "lyrics" and "coverImage" (two attachment types). Then each song `Token` can set them, referencing `attachmentType`. Why we need a contract-wide enumeration? To ensure type homogeniety for the UI.

Author: > data Source :: url(Proto, Dns, Id?) | urn(UrnPrefix, Id?) | storm(NodeAddr, Id?)

not sure I get what you mean by an `Id` here and why it is needed

Reviewer: By Id I mean an identifier such as a hash. A source would be a place the identifier could be requested from. That way an attachment could have a single identifier and there could be multiple sources outlined in a single place.

Author: From what I understand the source is a generic source for multiple items (tokens), and each token would be different. I.e. that can't be a "hash" of the source.

For instance, the source is a server URL ("https://token-provider.com/") and each token would specify their `media.uri` as `https://token-provider.com/some-nft-contract/token-hash1` etc. Sources ensure that all issuers uses only same-typed sources (especially important in case of decentralized issue)

Reviewer: The concern is if a source were go offline, it's important to be able to have multiple to search on. And a hash could make it content-addressable, so as long as the content is delivered and it matches the hash, it can be considered valid.

Author: The hash is a token-specific, not source-specific (the same source can be used by multiple tokens).

Having sources replaceable for the issued tokens will conflict with decentralized issue: a contract creator will now will be able to replace tokens issued by other parties onto something different.

Reviewer: They won't if it's content addressable. If the data hashes the same as the hash requested, redundancy can only improve reliability over time. Otherwise, NFTs will eventually go offline as sources inevitably fail.

Author: Sorry I am not getting it. It seems we are talking about different things.

I clearly distinguish RGB21 contract instance (which is a specific NFT contract issued under some of schema which implements RGB21 interface) from non-fungible token inside that contract.

`Sources` is a property of the contract, not a specific token under it.

Author: Please see my detailed explanations: https://github.com/LNP-BP/LNPBPs/pull/137#issuecomment-1360417616

interface RGB21
    global Name :: Nomination
    global Tokens :: [(Token, [Engraving])]
    global isFractional :: Bool

> 

Author: ```suggestion
global Reg :: Auxillary

```

Reviewer: `Auxillary` -> `Auxiliary`

Also, let's spell it out fully as `Registry`, because if left ambiguous, there's any number of things it could be confused for.

    owned Allocations+ :: Allocation
    owned IssueRight* :: Amount
    owned RenominationRight

> 

Author: ```suggestion
owned ControlRight

```

    -- returns information about known circulating supply
    read supplyKnown :: Amount
        count Self.state["Tokens"]
        -- sum Self.ops["issue"]..closed["usingRight"].state ?? 0

    -- maximum possible asset inflation
    read supplyLimit :: Amount
        sum Self.IssueRight..state ?? 0

    read isCirculationKnown :: Bool
        all Self.ops["issue"]..stateKnown

    op transfer    :: inputs [Allocation+] -> beneficiaries [Allocation]
                   !! -- options for operation failure:
                      inequalFractions(TokenId)
                    | nonFractionalToken

    -- question mark denotes optional operation, which may not be supported by 

>     -- some of schemata implementing the intrface

Reviewer: ```suggestion
    -- some of schemata implementing the interface
```

    op? engrave    :: Allocation -> Allocation
                   <- Engraving

    op? issue      :: IssueRight -> IssueRight, beneficiaries {Allocation}
                   <- tokens {Token}
                   !! invalidReserves(POR)

    -- decentralized issue

>     op? dcntrlIssue -> tokens {Token}, beneficiaries {Allocation}

Reviewer: ```suggestion
    op? decentralizedIssue -> tokens {Token}, beneficiaries {Allocation}
```
Let's call `dcntrlIssue` -> `decentralizedIssue`, since shorthand makes it less legible.

                    !! invalidReserves(POR)

> 

Author: ```suggestion

-- updates registry of auxillary data to a new version
op? update :: controlRight -> controlRight
           <- Reg

```

Reviewer: `auxillary` -> `auxiliary`
`Reg` -> `Registry`

Authors: Dr Maxim Orlovsky <orlovsky@lnp-bp.org>,
         Hunter Trujillo,
         Federico Tenga,
         Carlos Roldan,
         Olga Ukolova,
         Giacomo Zucco,
Comments-URI: <https://github.com/LNP-BP/LNPBPs/issues/70>
Status: Proposal
Type: Standards Track
Created: 2020-09-10
License: CC0-1.0
```

- [Abstract](#abstract)
- [Background](#background)
- [Motivation](#motivation)
- [Design](#design)
- [Specification](#specification)
- [Compatibility](#compatibility)
- [Rationale](#rationale)
- [Reference implementation](#reference-implementation)
- [Acknowledgements](#acknowledgements)
- [References](#references)
- [Copyright](#copyright)
- [Test vectors](#test-vectors)

## Abstract

## Background

## Motivation

## Design

- [x] Media as URI (which can be attachments distributed with Storm or usual URLs)
- [x] Small media ("previews" if given together with large media)
- [x] Fraction of assets
- [x] Engravings
- [x] Reserves
- [x] Possible decentralized issue

## Specification

Interface specification is the following Contractum code:

```haskell
-- # Defining main data structures

-- collectibles are usually scarse, so we limit their max number to 64k

> data ItemsCount :: U16

Author: ```suggestion
data ItemsCount :: U32
```

Reviewer 2: Understood, U32 looks like a fair trade-off.

-- each collectible item is unique and must have an id
data TokenId :: U16

-- Collectibles are owned in fractions of a single item; here the owned
-- fraction means `1/F`. Ownership of the full item is specified `F=1`;
-- half of an item as `F=2` etc.
data OwnedFraction :: U64

data TxOut :: txid [Byte ^ 32], vout U16

-- allocation of a single token or its fraction to some transaction output
data Allocation :: TokenId, OwnedFraction

data Media ::
    type MimeType,
    data [Byte]

data Attachment ::
    type MimeType,
    uri: Uri

data POR :: -- proof of reserves
    reserves TxOut,
    proof [Byte] -- schema-specific proof 

data Token ::
    id TokenId,
    name [Ascii ^ 1..40],
    details [Unicode ^ 40..256]?,
    preview Media?, -- always embedded preview media < 64kb
    media Attachment?, -- external media which is the main media for the token
    attachments { U8 ^ ..20 -> Attachment } -- auxillary attachments by type (up to 20 attachments)
    reserves POR? -- output containing locked bitcoins; how reserves are
                  -- proved is a matter of a specific schema implementation

> data Nomination :: 

Author: ```suggestion
data Nominal :: 
```

Reviewer: One thing, I think we should use the term Denomination here, as a reference to this:
https://en.wikipedia.org/wiki/Denomination_(currency)#Choice_of_name

The definition for the word Nominal has some adjacent meanings, but isn't as applicable. See:
https://www.merriam-webster.com/dictionary/nominal

I can have an English scholar provide his opinion if that helps.

Reviewer 2: As an alternative I think also something like "Identifier(s)" could fit.

Reviewer: @fedsten Not sure if you saw the email thread I included you two on, but Prof Andrew M Bailey suggested the same thing :)

Author: I think `Identifier` is too general and used in many other places, so leaving `Denomination`

    ticker [Ascii ^ 1..8],
    name [Ascii ^ 1..40],
    details [Unicode ^ 40..256]?,
    contract [Unicode]??,

data Auxiliary ::
    -- list of allowed sources
    sources [Source]
    
     -- each attachment type is a mapping from attachment id 
     -- (used as `Token.attachments` keys) to a short Ascii string
     -- verbally explaining the type of the attachment for the UI
     -- (like "sample" etc).
    attachmentTypes { U8 ^ ..32 -> [Ascii ^ 1..20] }

data Source :: url(Proto, Dns) | urn(UrnPrefix) | storm(NodeAddr)
data Proto :: http | https | httpxk | ws | wss | wssxk

interface RGB21

>     global Name :: Nomination

Author: ```suggestion
    global Name :: Nominal
```

    global Tokens :: [(Token, [Engraving])]
    global isFractional :: Bool

> global Registry :: Auxiliary

Author: ```suggestion
    global Registry :: Auxiliary
```

    owned Allocations+ :: Allocation
    owned IssueRight* :: Amount

>     owned RenominationRight

Author: ```suggestion
    owned NamingRight?
    
```

    read supplyKnown :: Amount
        count Self.state["Tokens"]
        -- sum Self.ops["issue"]..closed["usingRight"].state ?? 0

    -- maximum possible asset inflation
    read supplyLimit :: Amount
        sum Self.IssueRight..state ?? 0

    read isCirculationKnown :: Bool
        all Self.ops["issue"]..stateKnown

    op transfer    :: inputs [Allocation+] -> beneficiaries [Allocation]
                   !! -- options for operation failure:
                      inequalFractions(TokenId)
                    | nonFractionalToken

    -- question mark denotes optional operation, which may not be supported by 
    -- some of schemata implementing the interface

    op? engrave    :: Allocation -> Allocation
                   <- Engraving

    op? issue      :: IssueRight -> IssueRight, beneficiaries {Allocation}
                   <- tokens {Token}
                   !! invalidReserves(POR)

    -- decentralized issue
    op? decentralizedIssue -> tokens {Token}, beneficiaries {Allocation}
                    !! invalidReserves(POR)

-- updates registry of auxiliary data to a new version
op? update :: controlRight -> controlRight
           <- Registry

>     op? renominate :: RenominationRight -> RenominationRight

Author: ```suggestion
    op? rename :: NamingRight -> NamingRight? <- Nominal
```

```
LNPBP: 0021
Vertical: Smart contracts
Title: RGB non-fungible assets interface for collectibles (RGB-21)
Authors: Dr Maxim Orlovsky <orlovsky@lnp-bp.org>,
         Hunter Trujillo,

>          Federico Tenga,

Author: ```suggestion
         Federico Tenga,
         Zoe Faltibà,
```

- [Rationale](#rationale)
- [Reference implementation](#reference-implementation)
- [Acknowledgements](#acknowledgements)
- [References](#references)
- [Copyright](#copyright)
- [Test vectors](#test-vectors)

## Abstract

## Background

## Motivation

## Design

- [x] Media as URI (which can be attachments distributed with Storm or usual URLs)
- [x] Small media ("previews" if given together with large media)
- [x] Fraction of assets
- [x] Engravings
- [x] Reserves
- [x] Possible decentralized issue

## Specification

Interface specification is the following Contractum code:

```haskell
-- # Defining main data structures

-- collectibles are usually scarse, so we limit their max number to 64k
data ItemsCount :: U32

-- each collectible item is unique and must have an id
data TokenId :: U16

-- Collectibles are owned in fractions of a single item; here the owned
-- fraction means `1/F`. Ownership of the full item is specified `F=1`;
-- half of an item as `F=2` etc.
data OwnedFraction :: U64

data TxOut :: txid [Byte ^ 32], vout U16

-- allocation of a single token or its fraction to some transaction output
data Allocation :: TokenId, OwnedFraction

> 

Author: ```suggestion

data Engraving :: appliedTo TokenId, content Media

```

We forgot to add structure describing specific engraving

data Media ::
    type MimeType,
    data [Byte]

data Attachment ::
    type MimeType,

>     uri: Uri

Author: ```suggestion
    digest: [U8 ^ 32] -- this can be any type of 32-byte hash, like SHA256(d), BLACKE3 etc
```

data POR :: -- proof of reserves
    reserves TxOut,
    proof [Byte] -- schema-specific proof 

data Token ::
    id TokenId,
    name [Ascii ^ 1..40],
    details [Unicode ^ 40..256]?,
    preview Media?, -- always embedded preview media < 64kb
    media Attachment?, -- external media which is the main media for the token

>     attachments { U8 ^ ..20 -> Attachment } -- auxillary attachments by type (up to 20 attachments)

Author: ```suggestion
    attachments { U8 -> ^ ..20 Attachment } -- auxiliary attachments by type (up to 20 attachments)
```

`{ U8 ^ ..20 -> Attachment }` means `BTreeMap<[u8; 20]; Attachment>` with up to `u16::MAX` elements

`{ U8 -> ^ ..20 Attachment }` means `BTreeMap<u8; Attachment>` with up to 20 elements

    reserves POR? -- output containing locked bitcoins; how reserves are
                  -- proved is a matter of a specific schema implementation

data Denomination :: 
    ticker [Ascii ^ 1..8],
    name [Ascii ^ 1..40],
    details [Unicode ^ 40..256]?,
    contract [Unicode]??,

> data Auxiliary ::

Author: ```suggestion
```

Not needed anymore

>     -- list of allowed sources
>     sources [Source]
>     

Author: ```suggestion
```

>      -- each attachment type is a mapping from attachment id 
>      -- (used as `Token.attachments` keys) to a short Ascii string
>      -- verbally explaining the type of the attachment for the UI
>      -- (like "sample" etc).
>     attachmentTypes { U8 ^ ..32 -> [Ascii ^ 1..20] }

Author: ```suggestion
 -- each attachment type is a mapping from attachment id 
 -- (used as `Token.attachments` keys) to a short Ascii string
 -- verbally explaining the type of the attachment for the UI
 -- (like "sample" etc).
data AttachmentType :: id U8, description [Ascii ^ 1..20]
```

We move attachment types to the global state

> data Source :: url(Proto, Dns) | urn(UrnPrefix) | storm(NodeAddr)
> data Proto :: http | https | httpxk | ws | wss | wssxk

Author: ```suggestion
```

interface RGB21
    global Name :: Denomination

>     global Tokens :: [(Token, [Engraving])]

Author: ```suggestion
    global Tokens :: Token+
    global Engravings :: Engraving+
```

Each global state field in a contract operates either as a replacable single item - or as an append-only list of items - otherwise it is impossible to have a consensus  in client-side-validation. Thus, we introduce the same notation as with the owned state: `+` suffix for the state which can be defined multiple times; `?` for optional state, `*` for state which may be absent or present as a single item (and once defined can be never removed) or absence of suffix for the state which is always present and only can be replaced.

As for the engravings, due to this fact we need to separate them from the token definitions.

    global isFractional :: Bool

>     global Registry :: Auxiliary

Author: ```suggestion
    global AttachmentTypes :: AttachmentType+
```

    owned Allocations+ :: Allocation
    owned IssueRight* :: Amount
    owned DenominationRight?

>     
> owned ControlRight
> 

Author: ```suggestion
```

The list of an allowed attachment types should be fixed after issuance

    -- returns information about known circulating supply
    read supplyKnown :: Amount
        count Self.state["Tokens"]
        -- sum Self.ops["issue"]..closed["usingRight"].state ?? 0

    -- maximum possible asset inflation
    read supplyLimit :: Amount
        sum Self.IssueRight..state ?? 0

    read isCirculationKnown :: Bool
        all Self.ops["issue"]..stateKnown

    op transfer    :: inputs [Allocation+] -> beneficiaries [Allocation]
                   !! -- options for operation failure:
                      inequalFractions(TokenId)
                    | nonFractionalToken

    -- question mark denotes optional operation, which may not be supported by 
    -- some of schemata implementing the interface

    op? engrave    :: Allocation -> Allocation
                   <- Engraving

>     op? issue      :: IssueRight -> IssueRight, beneficiaries {Allocation}
>                    <- tokens {Token}
>                    !! invalidReserves(POR)

Author: ```suggestion
    op? issue      :: IssueRight -> IssueRight, beneficiaries {Allocation}
                   <- tokens {Token}, newAttachmentTypes {attachmentType}*
                   !! invalidReserves(POR)
```

    -- decentralized issue
    op? decentralizedIssue -> tokens {Token}, beneficiaries {Allocation}
                    !! invalidReserves(POR)

> -- updates registry of auxiliary data to a new version
> op? update :: controlRight -> controlRight
>            <- Registry
> 

Author: ```suggestion
```
The list of an allowed attachment types should be fixed after issuance
